Python/ast.c

```diff
@@ -657,10 +657,22 @@ new_identifier(const char *n, struct compiling *c)
             return NULL;
         }
         PyTuple_SET_ITEM(c->c_normalize_args, 1, id);
-        id2 = PyObject_Call(c->c_normalize, c->c_normalize_args, NULL);
+        /* Use _PyObject_FastCall() this way to conceal c->c_normalize_args
+           from the user. */
+        id2 = _PyObject_FastCall(c->c_normalize,
```

Comment:
Don't use `c->c_normalize_args`. Use just a 2-element C array.

Reply:
All right.
Would you mind to mention why this is better?

Reply:
The problem is with reusing the same tuple for passing arguments. If the fake `normalize()` save the reference to the tuple, it will see that an immutable tuple is mutated. It should be implemented in C, I can't reproduce the problem with Python code.

Just allocate a 2-element array on the stack, fill it with arguments, and pass it to the function. This will significantly simplify the code too.

Reply:
Or just revert you last change, I'll create a separate PR. These bugs are related to the same function, but can be fixed separately.

Reply:
I understand that we must conceal the tuple `c->c_normalize_args` from the user.
Doesn't my patch conceal it? I passed only the tuple items array to `_PyObject_FastCall()`, so even it doesn't have access to the tuple.
for example, `_PyObject_FastCall()` might eventually cause calling `function_code_fastcall()`, which would copy the args C array into `f_localsplus`, or it might call `_PyStack_AsTuple()`, which would copy the args C array into a new tuple.

Reply:
Or is my fix bad because it uses the internal structure of `tuple`?

Reply:
Yes, your patch conceal it. But using `c->c_normalize_args` as a buffer is suboptimal. You don't need a heap allocation, and the code would be simpler if use a stack variable.

Reply:
cool, thanks for the explanation :)